### PR TITLE
[pallet-revive] ensure the return data is reset if no frame was instantiated

### PR DESCRIPTION
**Pull request by xermicus**
Failed call frames do not produce new return data but still reset it.

### CODE REVIEW

**Comment by pgherveou** [comment]
> /cmd prdoc --audience runtime_dev --bump major

learning new command, does that generate the prdoc for you as well

**Comment by xermicus** [comment]
> > /cmd prdoc --audience runtime_dev --bump major
> 
> learning new command, does that generate the prdoc for you as well

[Indeed](https://github.com/paritytech/polkadot-sdk/pull/6045/commits/4f0f71a1468bedc734bfb69ac92cfbcf2960e686)! :tada:

File: substrate/frame/revive/src/exec.rs [hunk 1]
```diff
@@ -1378,6 +1382,10 @@ where
 	}
 
 	fn delegate_call(&mut self, code_hash: H256, input_data: Vec<u8>) -> Result<(), ExecError> {
+		// We reset the return data now, so it is cleared out even if no new frame was executed.
+		// This is for example the case for unknown code hashes or creating the frame fails.
+		*self.last_frame_output_mut() = Default::default();
```
Comment (pgherveou) [inline substrate/frame/revive/src/exec.rs:1387]: isn't it what we do here already

https://github.com/paritytech/polkadot-sdk/blob/4f0f71a1468bedc734bfb69ac92cfbcf2960e686/substrate/frame/revive/src/exec.rs#L993-L1003

need to double check the code, but maybe this is not needed overthere with these changes

Comment (pgherveou) [reply substrate/frame/revive/src/exec.rs:1387]: also nit maybe a clear_output would look nicer

Comment (xermicus) [reply substrate/frame/revive/src/exec.rs:1387]: The code in `fn run()` guarantees that the last saved output is empty upon running the new frame. If we want to make the caller responsible for this I can remove the code in `fn run()`.

Comment (pgherveou) [reply substrate/frame/revive/src/exec.rs:1387]: right we need both

**Review by pgherveou — APPROVED** [review]

File: substrate/frame/revive/src/exec.rs [hunk 2]
```diff
@@ -1331,14 +1331,18 @@ where
 		// is caught by it.
 		self.top_frame_mut().allows_reentry = allows_reentry;
 
-		let dest = T::AddressMapper::to_account_id(dest);
-		let value = value.try_into().map_err(|_| Error::<T>::BalanceConversionFailed)?;
```
[the remark this answers is not in the document]
Comment (athei) [reply substrate/frame/revive/src/exec.rs:1335]: Seems like a bug

File: prdoc/pr_6045.prdoc [hunk 3]
```diff
@@ -0,0 +1,8 @@
+title: '[pallet-revive] ensure the return data is reset if no frame was instantiated'
+doc:
+- audience:
+  - Runtime Dev
+  description: Failed call frames do not produce new return data but still reset it.
+crates:
+- name: pallet-revive
+  bump: major
```
Comment (athei) [inline prdoc/pr_6045.prdoc:8]: Shouldn't this be fix? No API changed.

Comment (xermicus) [reply prdoc/pr_6045.prdoc]: Changed to patch.

**Review by athei — APPROVED** [review]

**Pull request merged**